Domex neutron raytracing

CFGrote opened the pull request:
openPMD extension for neutron ray tracing simulations. https://github.com/PaNOSC-ViNYL/ViNYL-project/issues/15

*Implements issue:* #15 

## Description
Small extension to openPMD for neutron raytracing with the purpose of transferring neutron states between raytracing simulation packages.

Each neutron is described with an independent state with the following parameters:
- position (x, y, z)
- velocity (x, y, z)
- time
- weight
- polarization (optional) (x, y, z)

In addition there is support for a position and time offset, which are both optional.

The major change is the independent time for each neutron, which is useful as the neutrons are not interacting and thus can be propagated to different times in the simulation. The different times are useful in practice as the energy of the neutron can be deduced from the travel time to the sample, and it is thus common to propagate a number of rays to the sample and record position, velocity, time and weight.

The polarization is optional an optional record, but the components are required. This should be interpreted as the entire polarization record can be omitted, but if it isn't, all components are required.

## Affected Components

- `EXT-NRAYTRACING`

## Logic Changes

To my understanding the only required logic change is the possibility of having independent time for each neutron.

## Writer Changes

To my understanding no changes are required in the writer as a consequence of this update.

- `libopenPMD` (upcoming): https://github.com/ComputationalRadiationPhysics/libopenPMD/...

## Reader Changes

I do not see any required changes to readers.

- `openPMD-validator`: https://github.com/openPMD/openPMD-validator/...
- `openPMD-viewer`: https://github.com/openPMD/openPMD-viewer/...
- `yt`: https://github.com/yt-project/yt/...
- `VisIt`: https://github.com/openPMD/openPMD-visit-plugin
- `libopenPMD` (upcoming): https://github.com/ComputationalRadiationPhysics/libopenPMD/...
- `XDMF` (upcoming): https://github.com/openPMD/openPMD-tools/...

## Data Converter
No work done on data converters, as this is the first version of the neutron raytracing extension.

mads-bertelsen committed:
Added EXT_NRAYTRACING.md

mads-bertelsen committed:
First draft of neutron raytracing standard

Neutron raytracing standard with particle neutron state. Used EXT_BeamPhysics as base for values.

mads-bertelsen committed:
Addressed comments by Carsten

Split vectors into components (position, velocity, spin)
Added unitDimension and unitSI.

mads-bertelsen committed:
Note on coordinate convention

Mentioned that the beams nominal direction is the z direction.

CFGrote committed:
Remove redundant hierarchy level

CFGrote commented on EXT_NRAYTRACING.md:
```diff
@@ -0,0 +1,114 @@
+Introduction
+------------
+
+This extention is designed for transfer of neutron particle states between neutron raytracing simulation codes.
```
... for neutron raytracing data.

CFGrote commented on EXT_NRAYTRACING.md:
```diff
@@ -0,0 +1,114 @@
+
+This extention is designed for transfer of neutron particle states between neutron raytracing simulation codes.
+The neutrons are saved in rays with a given weight corresponding to expected intensity.
+This standard is not yet finished and is currently based on EXT_BeamPhysics / EXT_WAVEFRONT
```
can remove line 12

CFGrote commented on EXT_NRAYTRACING.md:
```diff
@@ -0,0 +1,114 @@
+The neutrons are saved in rays with a given weight corresponding to expected intensity.
+This standard is not yet finished and is currently based on EXT_BeamPhysics / EXT_WAVEFRONT
+
+Each record needs to have the same length, as the n'th index for each record correspond to particle n.
```
remove since not accurate anymore

mads-bertelsen replied:
Why is this not relevant anymore? Since for example position is stored as three separate lists, they have to be the same length to make sense.

CFGrote commented on EXT_NRAYTRACING.md:
```diff
@@ -0,0 +1,114 @@
+Particle Records
+----------------
+
+- `position/`
```
ok for me, but if you see good reasons to store all vector components in one set, please revert to the previous.

mads-bertelsen replied:
I think I have found a solution to my previous issue with optional vectors such as spin.

Current implementation:
- `spin/`
    - type: Optional 
    - description: Particle spin direction, no unit
    - `x` 
        - type: Optional *(real)*
        - description: x component of spin

Could be fixed by making the components required, but the entire spin group could remain optional. Would that work as expected?
- `spin/`
    - type: Optional 
    - description: Particle spin direction, no unit
    - `x` 
        - type: Required *(real)*
        - description: x component of spin

CFGrote commented on EXT_NRAYTRACING.md:
```diff
@@ -0,0 +1,114 @@
+        - unitDimension: [m] (1., 0., 0., 0., 0., 0., 0.)        
+        - unitSI: 1.0
+
+- `positionOffset/`
```
same as above comment on line 21

CFGrote commented on EXT_NRAYTRACING.md:
```diff
@@ -0,0 +1,114 @@
+        - unitDimension: [m] (1., 0., 0., 0., 0., 0., 0.)   
+        - unitSI: 1.0      
+    
+- `velocity/`
```
see above

CFGrote commented on EXT_NRAYTRACING.md:
```diff
@@ -0,0 +1,114 @@
+        - unitDimension: [m/s] (1., 0., -1.0, 0., 0., 0., 0.)
+        - unitSI: 1.0
+    
+- `spin/`
```
see above

CFGrote commented on EXT_NRAYTRACING.md:
```diff
@@ -0,0 +1,114 @@
+- `weight/`
+    - type: Required *(real)*
+    - description: Weight of the neutron ray with the physical unit of intensity, neutrons per second. When simulating a source with a given intensity, this intensity is split up into weights of the neutron rays that will be simulated.
+    - unitDimension: [unitless] (0., 0., 0., 1., 0., 0., 0., 0.)
```
if unitless, all components of `unitDimension` should be 0

CFGrote reviewed (commented):
Very good and clean!

If you get the chance, please try to add as much information as possible to the description of the merge request.
As mentioned, I'm open to revert to storing all vector components in one set. Just make it uniform over all vectors data (position, positionOffset, spin, ...).

Since I created the pull request, you will have to do the merge...

mads-bertelsen committed:
Fixed some errors in units

Some unit vectors had 8 entries instead of the 7 required.

willend committed:
Update EXT_NRAYTRACING.md

Suggested updates
* Use polarisation rather than spin
* UnitDimension for weight should be n/s

mads-bertelsen committed:
Merge pull request #7 from McStasMcXtrace/domex_neutron_raytracing

Update EXT_NRAYTRACING.md

mads-bertelsen committed:
Update EXT_NRAYTRACING.md

Updates to language mentioned by Carsten.
Reordered records to keep vectors together.
Changed polarization components to required, yet the entire polarization category is still optional, which I take means if the record is defined it must contain all components.
Further descriptions on what is required versus optional and length of each record.
